### PR TITLE
subclass windows in macos so they can be made resizable even with no decorations

### LINKED ISSUE #368: Can't resive/move windows without decorations (on macOS?)
see https://github.com/jwilm/alacritty/pull/947#issuecomment-352350723

### PR DESCRIPTION
**Pull request by sodiumjoe**
The subclassing logic was copied from servo's fork of glutin:
https://github.com/servo/glutin/blob/63026a0f4c8ba8bdc40787f8305ab008e5eb45e3/src/api/cocoa/mod.rs#L418

### CODE REVIEW

File: src/platform/macos/window.rs [hunk 1]
```diff
@@ -431,12 +431,21 @@ impl Window2 {
...
+                    NSWindowStyleMask::NSResizableWindowMask
             };
 
-            let window = IdRef::new(NSWindow::alloc(nil).initWithContentRect_styleMask_backing_defer_(
+            let window_superclass = Class::get("NSWindow").unwrap();
+            let mut decl = ClassDecl::new("WinitWindow", window_superclass).unwrap();
+            decl.add_method(sel!(canBecomeMainWindow), yes as extern fn(&Object, Sel) -> BOOL);
+            decl.add_method(sel!(canBecomeKeyWindow), yes as extern fn(&Object, Sel) -> BOOL);
+            decl.add_method(sel!(mouseDownCanMoveWindow), yes as extern fn(&Object, Sel) -> BOOL);
+            decl.add_method(sel!(isMovableByWindowBackground), yes as extern fn(&Object, Sel) -> BOOL);
```
Comment (paulrouget) [inline src/platform/macos/window.rs]: Are `mouseDownCanMoveWindow` and `isMovableByWindowBackground` necessary?

Does the window move when you click and drag anywhere in the window? If so, is that the behavior you want. And is this the behavior we want when there is a titlebar?

Comment (sodiumjoe) [reply src/platform/macos/window.rs]: Great point, and I also see that there's a separate platform attribute `movable_by_window_background` that already handles that independently. I'll remove those.

File: examples/window.rs [hunk 2]
```diff
@@ -1,10 +1,12 @@
 extern crate winit;
+use winit::os::macos::WindowBuilderExt;
 
 fn main() {
     let mut events_loop = winit::EventsLoop::new();
 
     let _window = winit::WindowBuilder::new()
         .with_title("A fantastic window!")
+        .with_decorations(false)
```
Comment (paulrouget) [inline examples/window.rs]: Not sure we want that in the example.

Comment (sodiumjoe) [reply examples/window.rs]: yup, that was just from my testing, I've removed it

**Comment by paulrouget** [comment]
The window is not movable at that point.

I'm not 100% sure what kind of behavior we want to achieve here. Like… how will drag-to-move work, or should we keep the window buttons, the rounded corner and the shadow.

But maybe adding these masks will make the window look better?

```
NSBorderlessWindowMask
NSTitledWindowMask
NSResizableWindowMask
NSFullSizeContentViewWindowMask
NSClosableWindowMask
NSMiniaturizableWindowMask
```

**Comment by sodiumjoe** [comment]
To be clear, this behavior is with and explicit `with_decorations(false)`. I have it running this way in a local build of alacritty, and it's exactly what I want. I think granular control of those masks makes more sense as platform attributes. I use a separate program to move/resize windows (hammerspoon), and I like the square corners and no border look.

> how will drag-to-move work

I've tested with

```rust
        .with_movable_by_window_background(true)
        .with_decorations(false)
```

and that creates an undecorated window that is movable by dragging anywhere in the window.

Maybe it makes sense for most or all of those separate masks to be granularly configurable as separate platform attribute methods, with `with_decorations(false)` resolving to some sane default preset combination of those methods?

I'd be happy to add that to this PR if that's the direction we want to go. I think it makes the configuration story for downstream projects a little complicated, though. E.g., currently alacritty just exposes a platform-agnostic boolean `decorations` field. I'm not sure how alacritty would expose these platform specific attributes through its config yaml file, or if @jwilm would event want to expose that level of granularity. I'll let him weigh in if he'd like.

**Comment by paulrouget** [comment]
So there are 3 different setups:

- titlebar + window controls + round borders
- window controls + round borders
- nothing + no round borders

winit only have one switch. So we need to pick between `window controls` or `nothing`.

Let's go for the current setup (nothing).

If anyone wants to add back the window controls, they can use `WindowExt::get_nswindow()` to reset the mask. In Alacritty, you could expose these 3 options.

**Comment by paulrouget** [comment]
You moved from NSFullSizeContentViewWindowMask to NSBorderlessWindowMask and removed the closable and miniaturizable masks. Can you explain exactly why these were necessary?

**Comment by sodiumjoe** [comment]
> NSFullSizeContentViewWindowMask to NSBorderlessWindowMask

My understanding of `NSFullSizeContentViewWindowMask` is that it's for [extending the content view into the title bar area](https://github.com/tomaka/winit/pull/389/files#diff-4c33d0eee5eebd4ec3689b0264753628R108). In my testing, it doesn't appear to have any effect when there is no title bar. The closable and miniaturizable masks also don't appear to have any effect when there's no title bar. If they were there for some reason I missed, I'd be happy to add them back.

**Comment by sodiumjoe** [comment]
So I think this PR achieves:

> nothing + no round borders

While https://github.com/tomaka/winit/pull/389 would allow:

> window controls + round borders

**Comment by paulrouget** [comment]
Ok. That works for me.

@tomaka happy with these changes?

**Comment by Osspial** [comment]
Are we sure this behavior is something we want implicitly true with the `with_decorations` flag? There are perfectly valid reasons to create an unresizable borderless window, such as with a right-click menu, and I don't want to lose the ability to do that on OSX. This is also inconsistent with other platforms - I know resizing a borderless window isn't allowed in Windows, at least right now.

**Comment by sodiumjoe** [comment]
@Osspial fair question. I guess it depends on what `decorations` means to wininit. I interpreted it to refer purely to the existence of visual elements, not behavior.

Also, iiuc, #389 would offer more granular controls over the masks such that you could still achieve non-resizeable borderless windows.

**Comment by paulrouget** [comment]
> Are we sure this behavior is something we want implicitly true 
>  I don't want to lose the ability to do that on OSX

Both scenario are covered. Both scenario are valid. It's a matter of what we do by default, what the expectation is.

I would recommend to land this as it is, and in #389 add enough documentation / examples to explain how to achieve both behaviors.

**Comment by sodiumjoe** [comment]
Anything blocking this from merging? @tomaka ?

**Comment by Osspial** [comment]
I'm okay with adding an option to make borderless windows resizable, but my biggest issue is that it both changes the current default behavior and makes the default behavior inconsistent across platforms. Also, I could be reading the diff incorrectly, but there doesn't seem to be a way to configure it to use the current behavior, which seems like an important thing to have.

**Comment by sodiumjoe** [comment]
My understanding is that borderless windows on windows and linux are resizable through other means, just not by dragging the borders (because the borders aren't there). Please correct me if I'm wrong.

If this is the case, then I think neither behavior is perfectly consistent across platforms. In fact, I'm not sure it's possible to achieve this in macos (i.e. borderless windows that are not resizeable by dragging the edges of the window, but resizeable through other means, like a window manager).

The more granular control is represented by #389, not this PR.

**Comment by Osspial** [comment]
Testing it, it looks like borderless windows on Windows are completely unresizable (or at least, I can't figure out a way to do it, even with keyboard shortcuts).

On Linux, it seems to depend on the window manager. I tested in with Xfce, and was able to resize borderless windows with `Alt+Right-Click`, but it still couldn't be resized with border dragging. Despite this, the behavior is still *mostly* consistent across platforms - borderless windows can't be resized by dragging on the edges.

**Comment by Ralith** [comment]
Note that under X11, it is technically impossible for a client to prevent the window manager from resizing its windows however it likes, so the existence of the possibility under some particular state shouldn't be read into too heavily.

**Comment by sodiumjoe** [comment]
`.with_decorations(false)` (current):

platform|resizable by dragging|resizable by window manager|borders
--------|---------------------|---------------------------|-------
windows|no|no|no
linux|no|yes|no
macos|no|no|no

`.with_decorations(false)` (proposed in this PR):

platform|resizable by dragging|resizable by window manager|borders
--------|---------------------|---------------------------|-------
windows|no|no|no
linux|no|yes|no
macos|yes|yes|no

@Osspial I can see your point about it being slightly more consistent behavior across
platforms. However, it still seems to me that it depends on what the meaning of
"decorations" is. I interpret it to refer purely to aesthetic concerns. For a
flag that concerns the behavior that you're talking about, I think a different
flag that refers specifically to the behavior would be more appropriate, e.g.
something like `.resizable(bool)`.

You're right that #389 doesn't have the settings for controlling this specifically, but I think it shows us how to use the window builder extension to expose all the granular settings that we would need to to allow it.

**Comment by sodiumjoe** [comment]
@tomaka would love your input on this. If this is not acceptable as is, I would like to get started on the WindowBuilderExt work to get the resizable window with no titlebar that I want.

**Comment by sodiumjoe** [comment]
@tomaka Hmmm, ok, so whom do I need to convince to get this merged or rejected?

**Comment by oschrenk** [comment]
I'm just a user but I sure do hope that this get's merged as I would love to have borderless windows that I can move around and resize (with external programs in macOS).

On that note @sodiumjoe If you could cut a release of your alacritty (with this change) that would be amazing.

**Comment by sodiumjoe** [comment]
@oschrenk you just need to clone alacritty, glutin, and winit, then reference your local copy of winit from glutin, and glutin from alacritty:

`glutin`'s `Cargo.toml`'s `[dependencies]` block:

```
winit = { path = "../winit" }
```

in `alacritty`'s `Cargo.toml`'s `[dependencies]` block:

```
glutin = { path = "../glutin" }
```

I've been using this build of alacritty since I started this PR and really enjoying it.

**Comment by paulrouget** [comment]
Maybe split this PR in 2. The `WinitWindow` subclass is necessary no matter what.

Then as what to do with the mask, it doesn't really matter, you can always do that via `WindowExt::get_nswindow()` I believe.

**Comment by sodiumjoe** [comment]
@paulrouget @Osspial @tomaka I've added a commit removing the resizable mask from an undecorated window. I believe that was the only remaining objection to this PR.

**Comment by paulrouget** [comment]
@sodiumjoe if you remove `NSResizableWindowMask`, you can't resize with the mouse anymore.

**Comment by paulrouget** [comment]
I'm still trying to understand what we want here… I'm confused.

**Comment by paulrouget** [comment]
Subclassing is necessary otherwise resizing is not possible, even by changing the mask via `WindowExt::get_nswindow()`. So, subclassing is 100% needed.

So let's have a PR just for that.

As for the default behavior, by adding the subclass, it makes the initial mask configuration work. So the window will be resizable *as initially intended*.

**Comment by sodiumjoe** [comment]
@paulrouget maybe I'm confused too, I believe the current state of the PR represents what you described, i.e. it adds the subclassing, but leaves resizability as it was before this PR, i.e. no `with_decorations(false)` windows are not resizable by any means. That last bit was the most recent commit.

The rest of the delta is just removing masks that have no effect on a window with no titlebar.

It sounds like you're disagreeing with @Osspial about whether a `with_decorations(false)` window should be resizable by default. I disagree too, but I'd rather just get this merged and get resizability by other means (`WindowExt::get_nswindow()` as you suggested).

I have a working build of alacritty doing exactly that with the following code:

```rust
        use glutin::os::macos::WindowExt;
        use self::cocoa::base::id;
        use window::cocoa::appkit::NSWindow;
        let nswindow = self.window.get_nswindow();
        unsafe {
            (nswindow as id).setStyleMask_((nswindow as id).styleMask() | cocoa::appkit::NSWindowStyleMask::NSResizableWindowMask);
        }
```

The unfortunate part of this is that it means dependents that want to use `get_nswindow()` also have to pull in `cocoa`.

My plan is to work on a PR along the lines of #389 to offer more granular control over the masks that a consumer can use. But it's not clear to me what that api should look like, how it will apply to other platforms, or how a consuming application like alacritty would expose those options to users. Maybe @jwilm can chime in about that part, so we can get an idea of what would be a good api for consumers of winit.

**Comment by paulrouget** [comment]
Going around in circles :) I'm happy with this change.

@tomaka this is an improvement and the initial behavior doesn't change, can we land this?

**Comment by sodiumjoe** [comment]
I made a PR to @robertgzr 's branch for #389 to add another `WindowBuilderExt` method to hide the titlebar, but keep resizability: https://github.com/robertgzr/winit/pull/1

But it depends on this one getting merged first.

**Comment by sodiumjoe** [comment]
@tomaka I believe all the objections to this PR have been resolved.

**Comment by sodiumjoe** [comment]
@tomaka anything else necessary here?

**Comment by tomaka** [comment]
Needs a rebase or a merge, and I'll merge!

**Comment by sodiumjoe** [comment]
@tomaka merged, thanks!

**Comment by mitchmindtree** [comment]
I just tried testing this on macos and this seems to break the `multiwindow.rs` example (I use this quite a lot downstream). Here's the backtrace:

```
   Compiling winit v0.11.2 (file:///Users/joshuabatty/Documents/Code/rust/winit)
    Finished dev [unoptimized + debuginfo] target(s) in 1.81 secs
     Running `target/debug/examples/multiwindow`
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:335:21
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:68
             at src/libstd/sys_common/backtrace.rs:57
             at src/libstd/panicking.rs:381
   2: std::panicking::default_hook
             at src/libstd/panicking.rs:397
   3: std::panicking::begin_panic
             at src/libstd/panicking.rs:577
   4: std::panicking::begin_panic
             at src/libstd/panicking.rs:538
   5: std::panicking::try::do_call
             at src/libstd/panicking.rs:522
   6: std::panicking::try::do_call
             at src/libstd/panicking.rs:498
   7: <core::ops::range::Range<Idx> as core::fmt::Debug>::fmt
             at src/libcore/panicking.rs:71
   8: <core::ops::range::Range<Idx> as core::fmt::Debug>::fmt
             at src/libcore/panicking.rs:51
   9: <core::option::Option<T>>::unwrap
             at /Users/travis/build/rust-lang/rust/src/libcore/macros.rs:20
  10: winit::platform::platform::window::Window2::create_window
             at src/platform/macos/window.rs:443
  11: winit::platform::platform::window::Window2::new
             at src/platform/macos/window.rs:321
  12: winit::platform::platform::Window::new
             at src/platform/macos/mod.rs:32
  13: <core::option::Option<T> as core::ops::try::Try>::into_result
             at src/window.rs:119
  14: winit::window::<impl winit::Window>::new
             at ./src/window.rs:135
  15: multiwindow::main
             at examples/multiwindow.rs:7
  16: std::rt::lang_start::{{closure}}
             at /Users/travis/build/rust-lang/rust/src/libstd/rt.rs:74
  17: std::panicking::try::do_call
             at src/libstd/rt.rs:59
             at src/libstd/panicking.rs:480
  18: panic_unwind::dwarf::eh::read_encoded_pointer
             at src/libpanic_unwind/lib.rs:101
  19: std::sys_common::bytestring::debug_fmt_bytestring
             at src/libstd/panicking.rs:459
             at src/libstd/panic.rs:365
             at src/libstd/rt.rs:58
  20: std::rt::lang_start
             at /Users/travis/build/rust-lang/rust/src/libstd/rt.rs:74
  21: multiwindow::main
```

**Comment by sodiumjoe** [comment]
I can repro, I'll check it out

**Comment by sodiumjoe** [comment]
@mitchmindtree fixed, can you verify?

**Comment by sodiumjoe** [comment]
@mitchmindtree awesome, thanks for the quick verification! Anything else I need to do to get this merged?

**Comment by sodiumjoe** [comment]
@tomaka just my weekly check-in on this PR 😄

**Comment by francesca64** [comment]
@sodiumjoe this should have a CHANGELOG entry.

**Comment by sodiumjoe** [comment]
@francesca64 done!

**Comment by francesca64** [comment]
1. The change should go under the Unreleased heading
2. Please try to better conform to the style of the rest of the CHANGELOG (i.e. don't add those additional line breaks, and in general)

**Comment by sodiumjoe** [comment]
Ok, how does it look now?

**Comment by francesca64** [comment]
@sodiumjoe Looks good, thanks!

While I'm able to merge this, I'm not sure how comfortable @tomaka is with me merging things. With how busy he is, I want to do as much to help out with winit as I can, so I'm hoping for some clarification on his policy.

...that said, since he already approved this, I'll merge it.

**Pull request merged**